django/db/models/sql/query.py

```diff
@@ -2339,6 +2339,9 @@ def set_group_by(self, allow_aliases=True):
             self.append_annotation_mask(group_by_annotations)
             self.select = tuple(values_select.values())
             self.values_select = tuple(values_select)
+            if self.selected is not None:
+                for index, value_select in enumerate(values_select):
+                    self.selected[value_select] = index
```

Comment:
`self.select` was re-organized without making sure `self.selected` was also kept in-sync which caused issues when grouping by `(annotation, col)` as `selected` was left as `{annotation: annotation, col: 1}` instead of `{[col](annotation: annotation, col: 0}` when `select` is set to `(col,)`.